### ISSUE TITLE
False error "Viewport Texture must be set to use it."

### ISSUE THREAD

**Issue by rarsp** [godotengine/godot#66247]
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

4.0 beta 1 mono

### System information

Windows 10

### Issue description

Scenes with viewport shows this errors even though viewport textures are set and working in materials:

ERROR: Viewport Texture must be set to use it.
   at: (scene/main/viewport.cpp:114)
ERROR: Viewport Texture must be set to use it.
   at: (scene/main/viewport.cpp:109)

### Steps to reproduce

Create simple scene with viewport and material that is using it as texture.

**Comment by Rindbee** [comment]
I guess the `ViewportTexture` is obtained by instantiating it instead of getting it from an existing `Viewport`. Some properties need to be set before it can be used. The error message is really not very detailed either.

**Comment by isaaccp** [comment]
I keep getting this too (in GDscript), but otherwise my SubViewport works fine.

**Comment by PedroPareja** [comment]
I got this same situation/error. The viewport texture does not work when exporting (it does work when executing from the editor)

**Comment by PedroPareja** [comment]
I found the solution here: https://github.com/godotengine/godot/issues/55471

I set the **subviewport update mode to always**. It seems **there's a bug with "update only when visible" option** and it has not been fixed

**Comment by Withaust** [comment]
In RC2 this error seems to be happening even with the update mode set to always...

**Comment by KettLovahr** [comment]
Yes, I'm having the same issue, no matter what configuration the SubViewport uses, it will act as if there's no Viewport set on debug or on the editor once the scene is reloaded.

**Comment by 1419427247** [comment]
![image](https://user-images.githubusercontent.com/33598732/222711631-668d554e-9cde-42aa-91f9-7353e406daf0.png)
V4.0. table. official [92bee43ad] Still the same

**Comment by baschermann** [comment]
### Godot version
4.0.1 RC2 mono

### System information
Windows 10

### Issue description
Scenes with viewport as a 3D texture shows errors on game startup. Mesh materials from viewport (2D) textures do not work and show up purple ingame, even though they show up in the editor.

E 0:00:00:0832   get_height: Viewport Texture must be set to use it.
  <C++ Error>    Condition "!vp" is true. Returning: 0
  <C++ Source>   scene/main/viewport.cpp:116 @ get_height()

These appear as errors, even though they work except for the mesh material option.

### Steps to reproduce
1. Create simple scene with viewport and 2D control node
2. Disable 3D in viewport
3. Use viewport as texture for material

Minimal reproduction project
[ViewportTexture66247.zip](https://github.com/godotengine/godot/files/11027571/ViewportTexture66247.zip)

![grafik](https://user-images.githubusercontent.com/13796592/226589046-69e78f8f-7b65-4102-878a-33ca98b87ff5.png)

**Comment by baschermann** [comment]
As a **_workaround_**, set the viewport texture in the parent node of both the SubViewport and the MeshInstance3D in code:

### Example in C#
```csharp
public override void _Ready()
{
      RenderingServer.FramePostDraw += OnFramePostDraw;
}

private void OnFramePostDraw()
{
      SubViewport subViewport = GetNode<SubViewport>("SubViewport");
      var viewportTexture = subViewport.GetTexture();

      MeshInstance3D meshInstance = GetNode<MeshInstance3D>("Mesh");

      StandardMaterial3D mat = new()
      {
          AlbedoTexture = viewportTexture
      };

      meshInstance.SetSurfaceOverrideMaterial(0, mat);
}
```

**Comment by michasng** [comment]
What seems to be happening is that this depends on the order of initialization. For me, the mesh was initialized first and it holds references to the uninitialized viewport. That issue resolves itself after a frame or so.
As mentioned, one workaround is to remove the viewport texture in the scene tree and create it programmatically.

**Comment by Rindbee** [comment]
> What seems to be happening is that this depends on the order of initialization. 

This has little to do with the order of instantiation. It mainly depends on whether `ViewportTexture::setup_local_to_scene()` is called and when it is called.

Currently, even if a resource has `local_to_scene` enabled, but its parent resource does not have `local_to_scene` enabled, the resource will not call `setup_local_to_scene()`.

**Comment by clemens-tolboom** [comment]
I followed along with the 3D https://docs.godotengine.org/en/stable/tutorials/animation/playing_videos.html#displaying-a-video-on-a-3d-surface on 4.0.3.rc1 ... my video plays well but errors when running from editor gives:
```
E 0:00:00:0580   get_width: Viewport Texture must be set to use it.
  <C++ Error>    Condition "!vp" is true. Returning: 0
  <C++ Source>   scene/main/viewport.cpp:111 @ get_width()

E 0:00:00:0580   get_height: Viewport Texture must be set to use it.
  <C++ Error>    Condition "!vp" is true. Returning: 0
  <C++ Source>   scene/main/viewport.cpp:116 @ get_height()

E 0:00:00:0623   setup_local_to_scene: Condition "proxy.is_valid()" is true.
  <C++ Source>   scene/main/viewport.cpp:89 @ setup_local_to_scene()
```

**Comment by ttencate** [comment]
Fixed by #75751?

**Comment by akien-mga** [comment]
> Fixed by #75751?

That's likely. Please test in 4.1 dev 2 or the upcoming 4.0.3 RC 2.

**Comment by Withaust** [comment]
4.1 dev 2 still has the issue with `get_height()` and `get_width()`
![image](https://github.com/godotengine/godot/assets/118038102/730d2f67-8b6f-4501-bec1-c360de2da752)
[Project.zip](https://github.com/godotengine/godot/files/11478873/Project.zip)

**Comment by Withaust** [comment]
Also, reopening the project greets you with `Viewport Texture must be set to use it.`, so it is not fixed at all
![image](https://github.com/godotengine/godot/assets/118038102/570212c0-a210-4bba-9118-f1221adb42f8)

**Comment by RichardR01** [comment]
This is still an issue in 4.02 stable. Having multiple errors, even though it technically works.
![10_viewport_errors](https://github.com/godotengine/godot/assets/46553425/00ae67aa-6495-4411-85e5-2e8f8d1a6f7a)

**Comment by FoggyFella** [comment]
@RichardR01 I think it has been fixed in 4.0.3 
https://github.com/godotengine/godot/pull/75751#issuecomment-1545520308

**Comment by RichardR01** [comment]
I upgraded to version 4.03 rc2. I am still having problems. Most of the errors are gone, but I still get 2 when I run the scene. Also the editor view isn't being updated, as it always shows white on my object. The object also doesn't get drawn in runtime, unless I add the viewport texture to the emission channel. It still technically works, but I think it is worse now that it was. The editor visual got updated when I reloaded godot. But the 2 errors till persist. This really needs to be tested thoroughly.
E 0:00:01:0410   get_height: Viewport Texture must be set to use it.
  <C++ Source>   scene/main/viewport.cpp:131 @ get_height()
E 0:00:01:0410   get_width: Viewport Texture must be set to use it.
  <C++ Source>   scene/main/viewport.cpp:121 @ get_width()

**Comment by Rindbee** [comment]
> Most of the errors are gone, but I still get 2 when I run the scene.

This happens during the scene instantiation phase. Resources are configured before nodes. The `get_width()`/`get_height()` of the texture is called when setting the texture for a material. See https://github.com/godotengine/godot/issues/66247#issuecomment-1548775331.

> Also the editor view isn't being updated, as it always shows white on my object.

This is a regression from #64388, see #77207.

> I am stilling getting the following error when loading up the scene file in the editor.
> `scene/main/viewport.cpp:185 - Condition "proxy.is_valid()" is true.`

This is caused by multiple calls to `setup_local_to_scene()` when the target viewport has not changed. As far as I know, toggling `resource_local_to_scene` causes `setup_local_to_scene()` to be called.

**Comment by DigitalN8m4r3** [comment]
same as the above and it really just kills my project since it depends on subviewports in a way that i cant even move forward without it, so yeah i hope this gets fixed soon

**Comment by clemens-tolboom** [comment]
@DigitalN8m4r3 in https://github.com/godotengine/godot/issues/66247#issuecomment-1483786200 there seems to be a workaround. If that works there is no killing. And what version you are working with?

**Comment by Delocop** [comment]
Getting the same error on 4.1 stable

**Comment by fire** [comment]
I hasty un-merged this in our fork. https://github.com/V-Sekai/godot/tree/viewport-path-update-fix.

The screen is pink and after the fix it works as it did in 4.0.

In 4.1 and 4.2 the screen is pink and the desktop copy of the vr viewport does not work. It's a blocker for us, so this is rush job fix.

**Comment by kapaseker** [comment]
Same Error 4.1.1

**Comment by Blue-Peanuts** [comment]
Error is gone for me in 4.2-dev5

**Comment by PITR-DEV** [comment]
Still present for me in 4.2 dev5

<img width="619" alt="image" src="https://github.com/godotengine/godot/assets/13086471/6b786672-b868-4f0d-b191-50db68de871c">

**Comment by SlashScreen** [comment]
regarding the [workaround](https://github.com/godotengine/godot/issues/66247#issuecomment-1483786200)
This works in 4.1.1. Gerre's a GDScript approximation:

```
extends MeshInstance3D


@export var viewport:SubViewport # Could instead get a child with an @onready


func _ready() -> void:
	RenderingServer.frame_post_draw.connect(on_draw.bind())


func on_draw() -> void:
	var tex = viewport.get_texture()
	var mat = StandardMaterial3D.new()
	mat.albedo_texture = tex
	set_surface_override_material(0, mat)
	RenderingServer.frame_post_draw.disconnect(on_draw.bind())
```

**Comment by 2fd5** [comment]
Still present in 4.2 beta 1

**Comment by chrisl8** [comment]
I built Godot from source using `master` today and I still get the error, so even in 4.2 Beta 3+
![image](https://github.com/godotengine/godot/assets/6188278/f4cd5d9c-fc9c-4c51-90d7-a1a7fe6f0997)

It all works fine in game though, so I guess it is a timing issue?

I will try some of the workarounds, as the error is distracting when I'm looking for real problems.

**Comment by mdarius31** [comment]
Are there plans to backport the fix to godot lts?

**Comment by Calinou** [comment]
> Are there plans to backport the fix to godot lts?

The issue isn't fuilly resolved in 4.x yet, so that needs to be done first before it can be backported to 3.x.

**Comment by nicemicro** [comment]
> The issue isn't fuilly resolved in 4.x yet as per the above comments, so that needs to be done first before it can be backported to 3.x.

Thank you! I confirm that a similar issue exists in 3.5.3.

**Comment by SoapSpangledGames** [comment]
> I found the solution here: #55471
> 
> I set the `SubViewport` **update mode to always** to fix the problem. It seems **there's a bug with "update only when visible" option** and it has not been fixed yet

My SubViewport update mode is set to Always, and I still get the error message when trying to duplicate the texture. I get no error when debugging in the editor, but it happens in the exported program (in my case, the Quest 3).

**Comment by SrColoma** [comment]
I found a solution while attempting to bypass that check, and it seems to be working fine. I want to recreate Prey's screens, and this is how it's turning out:

https://twitter.com/GalacticLake/status/1769937516889616637

Before, I was getting pink screens every time I started the project. This is my solution:
![image](https://github.com/godotengine/godot/assets/51213418/06971150-4db3-4a9c-9781-b0c4dfabf69a)

It seems the trick lies in creating an **exported variable** that stores a **packed scene** of what the subviewport will display. Inside the _ready function, **assign it as a child of the subviewport**, then make the necessary configurations, such as assigning the material and calling setup_local_to_scene(). Disabling the process mode isn't necessary; that's part of my multi-screen system. With this method, I no longer encounter errors because it's created and configured after the nodes are ready.

In the editor, the screens won't be visible, but they will appear in the game; that's the only issue.
![image](https://github.com/godotengine/godot/assets/51213418/e6cef6ad-88dc-4563-9895-b9b0f1b4ab1d)
![image](https://github.com/godotengine/godot/assets/51213418/46014e80-2a6d-496f-922d-ab3bdd604ac2)

**Comment by GeorgeS2019** [comment]
A demo to test the latest status of the Viewport

![Godot_v4 2 1-stable_mono_win64_JGIm0l0Sk4](https://github.com/godotengine/godot/assets/49812372/871cab1a-75af-468a-a3f8-fbcc181dbd70)
https://github.com/MASSHUU12/godot-control-3d/issues/3

**Comment by HeavensSword** [comment]
I was still running into this issue with Godot 4.3 Dev 5.

Based on some of the earlier explanations of the issue, it definitely seems to be an order of initialization issue. 
I had my SubViewport as a child to my MeshInstance3D previously, but I instead moved the SubViewport out as a sibling and placed it above the MeshInstance3D in the scene hierarchy which fixed the issue (no more errors).

Basically, based on how I understand Godot to work, children's Ready methods are called before their parent, but the opposite is true for EnterTree. I don't know exactly when the check occurs, but it's probably the latter. So by making the SubViewport a sibling and higher up in the hierarchy I believe it is initializing before the MeshInstance who wants to use its RenderTexture.

**Comment by chrisl8** [comment]
FWIW, the error, `Viewport Texture must be set to use it.` occurs in
 - v4.2.stable.official [46dc27791] (4.2.1 Stable release)
and
 - v4.3.dev.custom_build [9d6bdbc56] (latest build from Master at the moment)

When running the following Godot Demo projects:

[Procedural Materials](https://github.com/godotengine/godot-demo-projects/tree/master/3d/procedural_materials)
[GUI in 3D](https://github.com/godotengine/godot-demo-projects/tree/master/viewport/gui_in_3d)

So both of those demo projects can be used to test fixes for this issue in current and future versions.

**Comment by timkrief** [comment]
I'm using 4.2.2, my project is working fine but I have those issues, should we just ignore those warnings and errors for the time being?
![image](https://github.com/user-attachments/assets/62ed0340-6c02-40bb-aa21-c767cf989246)

**Comment by Withaust** [comment]
4.3 stable, still happening :(

**Comment by NervenCid** [comment]
I am using the 4.3 stable version, but i get the same error: 

This is the structure of my nodes: 

![image](https://github.com/user-attachments/assets/36ada80c-1e01-4885-99c0-e130038eccd6)

And i am defining the HUD as my texture on "MeshInstance3D" but i get the same errors:

![image](https://github.com/user-attachments/assets/4b29244d-5fd4-4b2d-bab2-8b884adbef5e)

How can i set a texture of a viewport? I can't see the option

**Comment by nkalpakis21** [comment]
i see this has been open for awhile. is there trouble getting a dev to fix this issue? what is blocking this for so long

**Comment by AThousandShips** [comment]
Please see the thread above for an explanation, but please don't bump issues without adding significant new information

**Comment by jrm780** [comment]
It's not clear to me whether this is supposed to be fixed. I see this issue was marked as completed with #97029, which was cherry-picked into 4.3-stable, but these errors and the magenta rendering in-game are still occurring in 4.3-stable.

**Comment by EduardoLemos567** [comment]
The bug still not fixed in version 4.5.1.
As we can see on this two scenes:
[Terrain.zip](https://github.com/user-attachments/files/23677711/Terrain.zip)

The first scene doesnt load the noise texture (cube is white), the second load correctly. 
Both raise the "ERROR: scene/main/viewport.cpp:188 - Viewport Texture must be set to use it." on the console.
Both use the same material marked as "Local to scene". 
The difference is the first one (the one that doesnt load the viewport texture) the material is attached into the mesh material slot.
The second is attached at the geometry material overwrite.

Should i open a new issue or this one can be reopen ?

**Comment by Rindbee** [comment]
>  Both use the same material marked as "Local to scene". 

For some use cases, this is insufficient. To be precise, **all** resources from the node to the `ViewportTexture` need to have `resource_local_to_scene` enabled.
